### PR TITLE
[SPARK-40925][SQL][SS] Fix stateful operator late record filtering

### PR DESCRIPTION
**Pull request by alex-balikov**
### What changes were proposed in this pull request?

This PR fixes the input late record filtering done by stateful operators to allow for chaining of stateful operators. Currently stateful operators are initialized with the current microbatch watermark and perform both input late record filtering and state eviction (e.g. producing aggregations) using the same watermark value. The state evicted (or aggregates produced) due to watermark advancing is behind the watermark and thus effectively late - if a following stateful operator consumes the output of the previous one, the input records will be filtered as late.

This PR provides two watermark values to the stateful operators - one from the previous microbatch to be used for late record filtering and the one from the current microbatch (as in the existing code) to be used for state eviction. This solves the above problem of the broken late record filtering.

Note that this PR still does not solve the issue of time-interval stream join producing records delayed against the watermark. Therefore time-interval streaming join followed by stateful operators is still not supported. That will be fixed in a follow up PR (and a SPIP) effectively replacing the single global watermark with conceptually watermarks per operator.

Also, the stateful operator chains unblocked by this PR (e.g. a chain of window aggregations) are still blocked by the unsupported operations checker. The new test for these scenarios - MultiStatefulOperatorsSuite has to explicitly disable the unsupported ops check. This again will be fixed in a follow-up PR.

### Why are the changes needed?

The PR allows Spark Structured Streaming to support chaining of stateful operators e.g. chaining of time window aggregations which is a meaningful streaming scenario.

### Does this PR introduce _any_ user-facing change?

With this PR, chains of stateful operators will be supported in Spark Structured Streaming.

### How was this patch tested?

Added a new test suite - MultiStatefulOperatorsSuite

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/MultiStatefulOperatorsSuite.scala [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+    q.processAllAvailable()
+    val progressWithData = q.recentProgress.lastOption.get
+    val stateOperators = progressWithData.stateOperators
+    assert(stateOperators.size === numTotalRows.size)
+    assert(stateOperators.map(_.numRowsTotal).toSeq === numTotalRows)
+    true
+  }
+
+  private def assertNumRowsDroppedByWatermark(
+    numRowsDroppedByWatermark: Seq[Long]): AssertOnQuery = AssertOnQuery { q =>
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/MultiStatefulOperatorsSuite.scala:387]: nit: 2 more spaces

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/MultiStatefulOperatorsSuite.scala [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+      // state does not get trimmed due to watermark advancement.
+      val dedupNoEventTime = aggStream
+        .dropDuplicates("count", "windowEnd")
+        .select(
+          $"windowEnd".cast("long").as[Long],
+          $"count".as[Long])
+
+      testStream(dedupNoEventTime)(
+        AddData(inputData, 1, 5, 10, 15),
+        CheckNewAnswer((5, 1), (10, 1), (15, 1)),
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/MultiStatefulOperatorsSuite.scala:353]: Shall we add the same level of trace comment here for W1 and W2?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/MultiStatefulOperatorsSuite.scala [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+        .withColumn("windowTimeMicros", expr("unix_micros(windowTime)"))
+        .dropDuplicates("count", "windowEnd", "windowTime")
+        .select(
+          $"windowEnd".cast("long").as[Long],
+          $"windowTimeMicros".cast("long").as[Long],
+          $"count".as[Long])
+
+      testStream(dedupWithEventTime)(
+        AddData(inputData, 1, 5, 10, 15),
+        CheckNewAnswer((5, 4999999, 1), (10, 9999999, 1), (15, 14999999, 1)),
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/MultiStatefulOperatorsSuite.scala:370]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala [hunk 4]
```diff
@@ -212,34 +212,71 @@ trait WatermarkSupport extends SparkPlan {
...
+  /** Predicate based on keys that matches data older than the late event filtering watermark */
+  lazy val watermarkPredicateForKeysForLateEvents: Option[BasePredicate] =
+    watermarkPredicateForKeys(watermarkExpressionForLateEvents)
+
+  /** Generate an expression that matches data older than the state eviction watermark */
+  lazy val watermarkPredicateForKeysForEviction: Option[BasePredicate] =
+    watermarkPredicateForKeys(watermarkExpressionForEviction)
+
+  private def watermarkPredicateForKeys(
+    watermarkExpression: Option[Expression]): Option[BasePredicate] = {
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala:250]: nit: 2 more spaces

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/MicroBatchWrite.scala [hunk 5]
```diff
@@ -28,7 +28,7 @@ import org.apache.spark.sql.connector.write.streaming.{StreamingDataWriterFactor
  */
 class MicroBatchWrite(epochId: Long, val writeSupport: StreamingWrite) extends BatchWrite {
   override def toString: String = {
-    s"MicroBathWrite[epoch: $epochId, writer: $writeSupport]"
+    s"MicroBatchWrite[epoch: $epochId, writer: $writeSupport]"
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/MicroBatchWrite.scala:31]: Nice catch :)

**Review by HeartSaVioR — COMMENTED** [review]
Looks OK in overall. Left comments.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SessionWindow.scala [hunk 6]
```diff
@@ -68,11 +68,29 @@ case class SessionWindow(timeColumn: Expression, gapDuration: Expression) extend
   with Unevaluable
   with NonSQLExpression {
 
+  private def inputTypeOnTimeColumn: AbstractDataType = {
```
Comment (HeartSaVioR) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SessionWindow.scala:71]: (Oh I missed that the change from rule side has already merged in as a part of introduction of `window_time`.)

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/MultiStatefulOperatorsSuite.scala [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+        // output: None
+        // state: None
+
+        // no-data batch triggered
+
+        // op1 W (21, 29)
+        // agg: None
+        // output: [20, 25) 5
+        // state: [25, 30) 4
+        // op2 W (20, 25)
```
[the remark this answers is not in the document]
Comment (alex-balikov) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/MultiStatefulOperatorsSuite.scala]: Good catch, thanks

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/MultiStatefulOperatorsSuite.scala [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+        // agg: [20, 30) (1, 5)
+        // output: None
+        // state: [20, 30) (1, 5)
+        CheckNewAnswer(),
+        assertNumStateRows(Seq(1, 1)),
+        assertNumRowsDroppedByWatermark(Seq(0, 0)),
+
+        // Move the watermark.
```
[the remark this answers is not in the document]
Comment (alex-balikov) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/MultiStatefulOperatorsSuite.scala:106]: Not sure what you mean - unfortunately there always will be state left as the watermark is up to the latest event time and that row will not be flushed. There are 3 microbatches in this test + 3 no-data microbatches. Why would another one help?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/MultiStatefulOperatorsSuite.scala [hunk 9]
```diff
@@ -0,0 +1,400 @@
...
+        .dropDuplicates("value", "eventTime")
+
+      val windowedAggregation = deduplication
+        .groupBy(window($"eventTime", "5 seconds") as 'window)
+        .agg(count("*") as 'count, sum("value") as 'sum)
+        .select($"window".getField("start").cast("long").as[Long],
+          $"count".as[Long])
+
+      testStream(windowedAggregation)(
+        // FIXME: we should revisit our watermark condition... we don't allow watermark to be
```
[the remark this answers is not in the document]
Comment (alex-balikov) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/MultiStatefulOperatorsSuite.scala]: https://issues.apache.org/jira/browse/SPARK-40942

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/MultiStatefulOperatorsSuite.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (HeartSaVioR) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/MultiStatefulOperatorsSuite.scala:106]: I meant adding a new batch (or existing batch) for testing late events.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 10]
```diff
@@ -1941,6 +1941,22 @@ object SQLConf {
...
       .createWithDefault(true)
 
+  val STATEFUL_OPERATOR_ALLOW_MULTIPLE =
+    buildConf("spark.sql.streaming.statefulOperator.allowMultiple")
+      .internal()
+      .doc("When true, multiple stateful operators are allowed to be present in a streaming " +
+        "pipeline. The support for multiple stateful operators introduces a minor (semantically " +
+        "correct) change in respect to late record filtering - late records are detected and " +
+        "filtered in respect to the watermark from the previous microbatch instead of the " +
+        "current one. This is a behavior change for Spark streaming pipelines and we allow " +
```
[the remark this answers is not in the document]
Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1951]: Yes, we will have a bunch of errors in test suites when we disable no-data batch. Most test cases are assuming that no-data batch always happens.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/MultiStatefulOperatorsSuite.scala (hunk 2 above)
Comment (alex-balikov) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/MultiStatefulOperatorsSuite.scala:374]: done

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/MultiStatefulOperatorsSuite.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (HeartSaVioR) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/MultiStatefulOperatorsSuite.scala:106]: Although not mandatory as other test case contains the verification. Just something "good to have".

**Review by HeartSaVioR — APPROVED** [review]
+1
I'll leave this PR in a couple of days to see whether we can get another eyes of review. I'll merge this in early next week there is no outstanding comment.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/MultiStatefulOperatorsSuite.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (alex-balikov) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/MultiStatefulOperatorsSuite.scala:106]: Good point. Added late records

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SessionWindow.scala (hunk 6 above)
Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SessionWindow.scala:71]: Nope. Just for future visibility.

**Pull request closed without merging**